pstoreds/addr_book.go

```diff
@@ -2,398 +2,338 @@ package pstoreds
+// datastore upon a miss, and returning a newly initialized record if the peer doesn't exist.
+//
+// loadRecord calls clean() on existing recordsrecord before returning it. If the record changes
+// as a result and the update argument is true, an async flush is queued.
+//
+// If the cache argument is true, the record is inserted in the cache when loaded from the datastore.
+func (ab *dsAddrBook) loadRecord(id peer.ID, cache bool, update bool) (pr *addrsRecord, err error) {
+	if e, ok := ab.cache.Get(id); ok {
+		pr = e.(*addrsRecord)
+		if pr.clean() && update {
+			ab.asyncFlush(pr)
 		}
+		return pr, nil
 	}
-	return nil
-}
 
-// dbInsert performs a transactional insert of the provided keys and values.
-func (mgr *dsAddrBook) dbInsert(keys []ds.Key, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) ([]bool, error) {
-	var (
-		err     error
-		existed = make([]bool, len(keys))
-		exp     = time.Now().Add(ttl)
-	)
+	pr = &addrsRecord{AddrBookRecord: &pb.AddrBookRecord{}}
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(id)))
+	data, err := ab.ds.Get(key)
 
-	txn, err := mgr.ds.NewTransaction(false)
-	if err != nil {
+	switch err {
+	case ds.ErrNotFound:
+		pr.Id = &pb.ProtoPeerID{ID: id}
+	case nil:
+		if err = pr.Unmarshal(data); err != nil {
+			return nil, err
+		}
+		if pr.clean() && update {
+			ab.asyncFlush(pr)
+		}
+	default:
 		return nil, err
 	}
-	defer txn.Discard()
 
-	ttltxn := txn.(ds.TTLDatastore)
-	for i, key := range keys {
-		// Check if the key existed previously.
-		if existed[i], err = ttltxn.Has(key); err != nil {
-			log.Errorf("transaction failed and aborted while checking key existence: %s, cause: %v", key.String(), err)
-			return nil, err
-		}
+	if cache {
+		ab.cache.Add(id, pr)
+	}
+	return pr, nil
+}
 
-		// The key embeds a hash of the value, so if it existed, we can safely skip the insert and
-		// just update the TTL.
-		if existed[i] {
-			switch mode {
-			case ttlOverride:
-				err = ttltxn.SetTTL(key, ttl)
-			case ttlExtend:
-				var curr time.Time
-				if curr, err = ttltxn.GetExpiration(key); err == nil && exp.After(curr) {
-					err = ttltxn.SetTTL(key, ttl)
+// flusher is a goroutine that takes care of persisting asynchronous flushes to the datastore.
+func (ab *dsAddrBook) flusher() {
+	ab.done.Add(1)
+	defer ab.done.Done()
+
+	for {
+		select {
+		case fj := <-ab.flushJobCh:
+			if cached, ok := ab.cache.Peek(fj.Id.ID); ok {
+				// Only continue flushing if the record we have in memory is the same as for which the flush
+				// job was requested. If it's not in memory, it has been evicted and we don't know if we hold
+				// the latest state or not. Similarly, if it's cached but the pointer is different, it means
+				// it was evicted and has been reloaded, so we're also uncertain if we hold the latest state.
+				if pr := cached.(*addrsRecord); pr != fj {
+					pr.RLock()
+					pr.flush(ab.ds)
```

Comment:
Flush modifies `pr.dirty` so we should either take full lock or move locking to `pr.flush`.

Reply:
Well spotted. In an earlier version the record didn't have a `dirty` flag, so flushing was only a read operation. Fixed.